File: scripts/reset_db.js
#!/usr/bin/env node
const { Client } = require("pg");
const readline = require("readline");

const DB_URL = process.env.DATABASE_URL ||
  "postgresql://neondb_owner:npg_Zd4mRIq9PCjk@ep-square-hill-b3y1avud-pooler.c-4.ap-southeast-1.aws.neon.tech/neondb?sslmode=require";

const TABLES = [
  "tg_hourly_limits","tg_chat_sessions","community_messages",
  "chat_messages","chat_sessions","alerts","reports",
  "scans","monitoring","password_checks","ppt_rate_limits",
  "monitoring_history","analyses","messages","sessions","users",
];

const SCHEMA = `
CREATE EXTENSION IF NOT EXISTS "pgcrypto";

CREATE TABLE users (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  name VARCHAR(255), email VARCHAR(255) UNIQUE NOT NULL,
  password_hash VARCHAR(255), firebase_uid VARCHAR(255) UNIQUE,
  telegram_id BIGINT UNIQUE, telegram_username VARCHAR(255),
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE monitoring (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  target_url TEXT NOT NULL, status VARCHAR(50) DEFAULT 'active',
  check_type VARCHAR(50) DEFAULT 'full',
  schedule VARCHAR(100), schedule_type VARCHAR(50),
  schedule_interval_minutes INTEGER,
  schedule_hour INTEGER DEFAULT 0, schedule_minute INTEGER DEFAULT 0,
  schedule_day INTEGER DEFAULT 0, schedule_month INTEGER DEFAULT 0,
  schedule_weekday INTEGER DEFAULT 1,
  last_run TIMESTAMPTZ, next_run TIMESTAMPTZ, updated_at TIMESTAMPTZ DEFAULT NOW(),
  previous_result JSONB DEFAULT '{}',
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE scans (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  target_url TEXT NOT NULL,
  raw_data JSONB DEFAULT '{}',
  findings JSONB DEFAULT '[]', technologies JSONB DEFAULT '[]',
  headers JSONB DEFAULT '{}', security_headers JSONB DEFAULT '{}',
  tls_info JSONB DEFAULT '{}', cookies JSONB DEFAULT '[]',
  redirects JSONB DEFAULT '[]', dns_info JSONB DEFAULT '{}',
  tls_check_info JSONB DEFAULT '{}',
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE reports (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  scan_id UUID, target_url TEXT, report_type VARCHAR(50),
  title TEXT, content JSONB DEFAULT '{}',
  summary TEXT, risk_level VARCHAR(50),
  version INTEGER DEFAULT 1, parent_report_id UUID,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE alerts (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  monitoring_id INTEGER REFERENCES monitoring(id) ON DELETE CASCADE,
  alert_type VARCHAR(50), message TEXT,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE chat_sessions (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  title VARCHAR(255) DEFAULT 'New Session',
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE chat_messages (
  id SERIAL PRIMARY KEY,
  session_id INTEGER REFERENCES chat_sessions(id) ON DELETE CASCADE,
  role VARCHAR(50) NOT NULL, content TEXT NOT NULL,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE sessions (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  title TEXT, created_at TIMESTAMPTZ DEFAULT NOW(),
  updated_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE messages (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  session_id UUID REFERENCES sessions(id) ON DELETE CASCADE,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  role TEXT, content TEXT, metadata JSONB DEFAULT '{}',
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE analyses (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  scan_id UUID, user_id UUID, target_url TEXT, summary TEXT,
  findings_analyzed JSONB DEFAULT '[]', correlations JSONB DEFAULT '[]',
  information_leakage JSONB DEFAULT '[]', improvements JSONB DEFAULT '[]',
  severity_summary JSONB DEFAULT '{}', risk_explanation TEXT,
  raw_analysis TEXT, model_used TEXT, tokens_used INTEGER,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE monitoring_history (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  monitor_id INTEGER, user_id UUID,
  scan_data JSONB DEFAULT '{}', dns_data JSONB DEFAULT '{}',
  tls_data JSONB DEFAULT '{}', changes_detected JSONB DEFAULT '[]',
  status TEXT, created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE password_checks (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id UUID, strength_score INTEGER, crack_time TEXT,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE ppt_rate_limits (
  id SERIAL PRIMARY KEY,
  user_id UUID, export_date DATE DEFAULT CURRENT_DATE,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE tg_chat_sessions (
  id SERIAL PRIMARY KEY,
  telegram_id BIGINT UNIQUE NOT NULL,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  is_active BOOLEAN DEFAULT TRUE,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
CREATE TABLE tg_hourly_limits (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  hour_start TIMESTAMPTZ NOT NULL,
  ai_chat_count INTEGER DEFAULT 0, report_count INTEGER DEFAULT 0,
  scan_count INTEGER DEFAULT 0, UNIQUE(user_id, hour_start)
);
CREATE TABLE community_messages (
  id SERIAL PRIMARY KEY,
  user_id UUID REFERENCES users(id) ON DELETE CASCADE,
  user_name VARCHAR(255), message TEXT NOT NULL,
  created_at TIMESTAMPTZ DEFAULT NOW()
);
`;

const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
function ask(q) { return new Promise(r => rl.question(q, r)); }

async function main() {
  console.log("NESTI DATABASE RESET (Full Schema)\n");
  const a = await ask("Type 'yes' to reset: ");
  if (a.trim().toLowerCase() !== "yes") { console.log("Cancelled."); rl.close(); return; }
  const c = new Client({ connectionString: DB_URL });
  await c.connect();
  try {
    for (const t of TABLES) { await c.query(`DROP TABLE IF EXISTS ${t} CASCADE`); console.log(`  Dropped: ${t}`); }
    await c.query(SCHEMA);
    console.log("\nAll tables recreated with full schema!");
  } catch (e) { console.error("Error:", e.message); }
  finally { await c.end(); rl.close(); }
}
main();

File: scripts/reset_db.py
#!/usr/bin/env python3
"""
NESTI Dynamic Database Reset
Auto-discovers all tables, dumps schema, drops and recreates.
No hardcoded table list.
"""
import os
import psycopg2

DB_URL = os.getenv(
    "DATABASE_URL",
    "postgresql://neondb_owner:npg_Zd4mRIq9PCjk@ep-square-hill-b3y1avud-pooler.c-4.ap-southeast-1.aws.neon.tech/neondb?sslmode=require",
)


def get_all_tables(cur):
    cur.execute(
        "SELECT table_name FROM information_schema.tables "
        "WHERE table_schema = 'public' ORDER BY table_name"
    )
    return [r[0] for r in cur.fetchall()]


def build_create_sql(cur, tables):
    """Build CREATE TABLE statements, split into base (no FK) and FK adds."""
    base_stmts = []
    fk_stmts = []

    for table in tables:
        cur.execute("""
            SELECT column_name, data_type, character_maximum_length,
                   is_nullable, column_default
            FROM information_schema.columns
            WHERE table_name = %s AND table_schema = 'public'
            ORDER BY ordinal_position
        """, (table,))
        columns = cur.fetchall()

        cur.execute("""
            SELECT tc.constraint_name, tc.constraint_type,
                   kcu.column_name, ccu.table_name AS foreign_table,
                   ccu.column_name AS foreign_column
            FROM information_schema.table_constraints tc
            JOIN information_schema.key_column_usage kcu
              ON tc.constraint_name = kcu.constraint_name
            LEFT JOIN information_schema.constraint_column_usage ccu
              ON tc.constraint_name = ccu.constraint_name
            WHERE tc.table_name = %s AND tc.table_schema = 'public'
        """, (table,))
        constraints = cur.fetchall()

        col_lines = []
        pk_cols = []

        for col_name, data_type, max_len, nullable, default in columns:
            is_serial = default and 'nextval(' in str(default)
            col_def = f'  "{col_name}"'
            if is_serial:
                col_def += ' SERIAL'
            else:
                col_def += f' {data_type.upper()}'
                if max_len:
                    col_def += f'({max_len})'
                if nullable == 'NO':
                    col_def += ' NOT NULL'
                if default and default != 'NULL' and not is_serial:
                    if 'gen_random_uuid()' in str(default):
                        col_def += ' DEFAULT gen_random_uuid()'
                    elif 'now()' in str(default).lower():
                        col_def += ' DEFAULT NOW()'
                    elif 'current_date' in str(default).lower():
                        col_def += ' DEFAULT CURRENT_DATE'
                    elif 'current_timestamp' in str(default).lower():
                        col_def += ' CURRENT_TIMESTAMP'
                    elif default == 'true':
                        col_def += ' DEFAULT TRUE'
                    elif default == 'false':
                        col_def += ' DEFAULT FALSE'
                    elif default.startswith("'") or default.startswith('"'):
                        col_def += f' DEFAULT {default}'
                    else:
                        try:
                            float(default)
                            col_def += f' DEFAULT {default}'
                        except (ValueError, TypeError):
                            col_def += f" DEFAULT '{default}'"
            col_lines.append(col_def)

        # PK and UNIQUE
        seen_unique = set()
        for cname, ctype, ccol, ftable, fcol in constraints:
            if ctype == 'PRIMARY KEY' and ccol:
                pk_cols.append(ccol)
            elif ctype == 'UNIQUE' and ccol and ccol not in pk_cols and ccol not in seen_unique:
                seen_unique.add(ccol)
                col_lines.append(f'  UNIQUE("{ccol}")')

        if pk_cols:
            pk_str = ', '.join(f'"{c}"' for c in pk_cols)
            col_lines.append(f'  PRIMARY KEY ({pk_str})')

        base_stmts.append(f'CREATE TABLE {table} (\n' + ',\n'.join(col_lines) + '\n);')

        # Collect FKs
        for cname, ctype, ccol, ftable, fcol in constraints:
            if ctype == 'FOREIGN KEY' and ftable and fcol:
                fk_stmts.append(
                    f'ALTER TABLE {table} ADD CONSTRAINT {cname} '
                    f'FOREIGN KEY("{ccol}") REFERENCES {ftable}("{fcol}") ON DELETE CASCADE;'
                )

    return base_stmts, fk_stmts


def main():
    print("========================================")
    print("  NESTI DYNAMIC DATABASE RESET")
    print("========================================\n")

    confirm = input("Type 'yes' to reset ALL tables: ").strip().lower()
    if confirm != "yes":
        print("Cancelled.")
        return

    conn = psycopg2.connect(DB_URL)
    cur = conn.cursor()

    # Discover
    tables = get_all_tables(cur)
    print(f"Found {len(tables)} tables\n")

    # Dump schema
    print("Dumping schema...")
    base_stmts, fk_stmts = build_create_sql(cur, tables)

    # Drop all
    print("Dropping all tables...")
    cur.execute("SELECT table_name FROM information_schema.tables WHERE table_schema = 'public'")
    all_tables = [r[0] for r in cur.fetchall()]
    for t in all_tables:
        cur.execute(f"DROP TABLE IF EXISTS {t} CASCADE")
        print(f"  Dropped: {t}")

    # Recreate base tables (no FKs)
    print("\nRecreating tables...")
    for stmt in base_stmts:
        cur.execute(stmt)

    # Add FKs
    print("Adding foreign keys...")
    for stmt in fk_stmts:
        cur.execute(stmt)

    # Post-reset fixups: constraints that dynamic dump may miss
    print("Applying fixups...")
    fixups = [
        # tg_hourly_limits needs composite UNIQUE for ON CONFLICT
        "ALTER TABLE tg_hourly_limits DROP CONSTRAINT IF EXISTS tg_hourly_limits_user_id_key",
        "ALTER TABLE tg_hourly_limits DROP CONSTRAINT IF EXISTS tg_hourly_limits_hour_start_key",
        "ALTER TABLE tg_hourly_limits ADD CONSTRAINT tg_hourly_limits_user_hour UNIQUE (user_id, hour_start)",
    ]
    for sql in fixups:
        try:
            cur.execute(sql)
        except Exception:
            pass

    conn.commit()
    conn.close()

    print("\n========================================")
    print(f"  DONE! {len(tables)} tables recreated")
    print("========================================")


if __name__ == "__main__":
    main()
